src/parallel.cpp

```diff
@@ -549,6 +549,10 @@ void HandleBlockMessageThread(CNode *pfrom, const string strCommand, const CBloc
             if (nDoS > 0)
                 dosMan.Misbehaving(pfrom, nDoS);
         }
+
+        // the current fork is bad due to this block so reset the best header to the best fully-validated block
+        // so we can download another fork of headers (and blocks).
+        pindexBestHeader = chainActive.Tip();
```

Comment:
@ptschip, take a look at main.cpp:SendMessages() lines 7380 to 7413, and in particular line 7400:
                // BU Bug fix for Core:  Don't start downloading headers unless our chain is shorter
                if (pindexStart->nHeight < pto->nStartingHeight)

This chunk of code only downloads from nodes that have longer headers than our pindexBestHeader.  But what if our pindexBestHeader is pointing to a blockchain minority fork that we have eliminated due to the "invalidateblock" RPC call or because of the BUIP055 forking logic?

In that case we will NEVER download from a node that is presenting the minority fork, because the (invalid) majority fork is longer.

The above code resets the pindexBestHeader to the best known-good block when it finds a fork situation (basically when it finds a situation where headers are good but blocks are bad).  Is there ever any situation where we want pindexBestHeader to point to the best header even if its blocks could be bad?

I thought no, so I made this more comprehensive fix.  But I could just as easily fix my problem by not touching pindexBestHeader at all.  Instead I'd use chainActive.Tip()->nHeight on line 7400 rather than pindexBestHeader->nHeight...

What are your thoughts on this?